// general-framework/lib/window/WindowService.js
import { WindowEvents } from './WindowEvents';
import { EventDispatcher } from '../EventDispatcher';
export class WindowService {
    /**
     * Inits window service
     */
    static init() {
        if (document.readyState === 'complete' || document.readyState === 'interactive') {
            this.dispatcher.dispatch(WindowEvents.DOM_ELEMENT_LOADED);
        }
        else {
            document.addEventListener('DOMContentLoaded', () => this.dispatcher.dispatch(WindowEvents.DOM_ELEMENT_LOADED));
        }
        const resize = () => this.dispatcher.dispatch(WindowEvents.RESIZE, {
            width: this.getWidth(),
            height: this.getHeight()
        });
        resize();
        window.addEventListener("resize", () => setTimeout(resize, 250));
    }
    /**
     * Enables full screen
     */
    static enableFullScreen() {
        const elem = document.documentElement;
        if (!this.isFullScreen) {
            if (elem.requestFullscreen) {
                elem.requestFullscreen();
            }
            else if (elem.mozRequestFullScreen) { /* Firefox */
                elem.mozRequestFullScreen();
            }
            else if (elem.webkitRequestFullscreen) { /* Chrome, Safari and Opera */
                elem.webkitRequestFullscreen();
            }
            else if (elem.msRequestFullscreen) { /* IE/Edge */
                elem.msRequestFullscreen();
            }
        }
    }
    /**
     * Disables full screen
     */
    static disableFullScreen() {
    }
    /**
     * Gets whether is full screen
     */
    static get isFullScreen() {
        return (document.fullscreenElement && document.fullscreenElement !== null) || (document.mozFullScreen || document.webkitIsFullScreen);
    }
    /**
     * Gets window width
     * @returns width
     */
    static getWidth() {
        return window.innerWidth;
    }
    /**
     * Gets window height
     * @returns height
     */
    static getHeight() {
        return window.innerHeight;
    }
}
WindowService.dispatcher = EventDispatcher.getInstance();
