// general-framework/lib/modalPopup/util/getModalPopupStyles.js
const defaultStyles = {
    background: 'rgba(255, 255, 255, .73)',
    color: '#4a4a4a',
    fontStyle: 'initial',
    btnFontColor: '#fff',
    primaryBackground: '#3664a9',
    successBackground: '#438c4a',
    dangerBackground: '#4a4a4a',
    warningBackground: '#c1bd1b',
    bgrBorderRadius: '10px',
    btnBorderRadius: '8px',
    btnFontSize: '20px',
};
/**
 * Bring custom styles to inject them in project
 * @param {IModalPopupStyles} customStyles - styles from project to customize popup
 * @returns {string}
 */
export const getModalPopupStyles = (customStyles = {}) => {
    // aggregate default styles with custom
    const styles = Object.assign({}, defaultStyles, customStyles);
    return (`
        #popup {
            display: none;
            position: fixed;
            top: 0;
            left: 0;
            z-index: 1000;
            width: 100%;
            height: 100%;
            background: rgba(0, 0, 0, .8);
        }

        #popup.active {
            display: block;
        }

        #popup .wrapper {
            position: relative;
            display: flex;
            justify-content: space-between;
            flex-direction: column;
            align-items: center;
            width: 100%;
            max-width: 650px;
            min-height: 356px;
            margin: auto;
            padding: 3% 1% 0;
            top: 50%;
            transform: translateY(-50%);
            background: ${styles.background};
            box-shadow: 0 0 20px rgba(255,255,255,.2);
            text-align: center;
            color: ${styles.color};
            font-size: 26px;
            line-height: 47px;
            font-family: arial, sans-serif;
            font-style: ${styles.fontStyle};
            border-radius: ${styles.bgrBorderRadius};
        }

        #popup .messages {
            height: 100%;
            display: flex;
            flex-direction: column;
            justify-content: center;
        }

        #popup .header {
            order: 0;
            margin-top: 0px;
            text-transform:uppercase;
            font-weight:bold;
        }

        #popup .body {
            order: 1;
            margin:20px;
        }

        #popup p {
            margin: 0;
            display:block;
        }

        #popup .buttons {
            justify-content: center;
            order: 10;
            width: 100%;
            margin: 4% 0;
            display: flex;
            align-items: center;
        }

        #popup button {
            margin: 0 6px;
            width: 30%;
            padding: 13px 0;
            color: ${styles.btnFontColor};
            border-radius: ${styles.btnBorderRadius};
            border: 0;
            font-family: arial, sans-serif;
            font-size: ${styles.btnFontSize};
            line-height: 26px;
            font-weight: bold;
            text-transform: uppercase;
            user-select: none;
            cursor: pointer;
            outline: none;
        }
        #popup .primary {
            background: ${styles.primaryBackground};
        }
        #popup .success {
            background: ${styles.successBackground};
        }
        #popup .danger {
            background: ${styles.dangerBackground};
            color: #ccc;
        }
        #popup .warning {
            background: ${styles.warningBackground};
        }

        #popup .close-wrap {
            display: flex;
            align-items: center;
            justify-content: center;
            position: absolute;
            width: 30px;
            height: 30px;
            right: 10px;
            top: 10px;
            cursor: pointer;
            border-radius: 6px;
        }
        
        #popup .close-wrap:hover {
            background: #333333;
        }
        #popup .close-wrap:active {
            background: #483D00;
        }
        #popup .btn-close {
            width: 19px;
            height: 19px;
            transform: rotate(45deg);
            cursor: pointer;
        }

        #popup .btn-close::before,
        #popup .btn-close::after {
            content: '';
            position: absolute;
            background: #c4c4c4;
            border-radius: 2px;
            left: 0;
            top: 0;
        }

        #popup .btn-close::before {
            top: 8px;
            width: 100%;
            height: 2px;
        }
        #popup .btn-close::after {
            left: 8px;
            width: 2px;
            height: 100%;
        }
        @media screen and (max-height: 400px) {
            #popup .wrapper {
                height: 100%;
                min-height: 0;
                max-width: none;
                border-radius: none;
            }
            #popup p {
                font-size: 7vh;
                line-height: 1.5;
            }
        }
    `);
};
