// general-framework/src/reality-check/RealityCheckModel.ts
/**
 * Save and return values for reality checking
 */

export class RealityCheckModel {
    private static _time_elapsed: number;
    private static _time_limit: number;
    private static _exit_URL: string;
    private static _lobby_URL: string;
    private static _lang: string;
    private static _isFrozen: boolean;
    private static _isShowHomeButton: boolean;
    private static _startTime: number;
    private static _history_URL: string;
    private static _rcInfoUrl: string;

    public static get time_elapsed(): number {
        return this._time_elapsed ? this._time_elapsed : 0;
    }

    public static set time_elapsed(value: number) {
        this._time_elapsed = value;
    }

    public static get time_limit(): number | null {
        return this._time_limit ? this._time_limit : null;
    }

    public static set time_limit(value: number) {
        this._time_limit = value;
    }

    public static get exit_URL(): string {
        return this._exit_URL;
    }

    public static set exit_URL(value: string) {
        this._exit_URL = value;
    }

    public static get lang(): string {
        return this._lang;
    }

    public static set lang(value: string) {
        this._lang = value;
    }

    public static get isFrozen(): boolean {
        return this._isFrozen;
    }

    public static set isFrozen(value: boolean) {
        this._isFrozen = value;
    }

    public static get startTime(): number {
        return this._startTime;
    }

    public static set startTime(value: number) {
        this._startTime = value;
    }

    static get lobby_URL(): string {
        return this._lobby_URL;
    }

    static set lobby_URL(value: string) {
        this._lobby_URL = value;
    }

    static get history_URL(): string {
        return this._history_URL;
    }

    static set history_URL(value: string) {
        this._history_URL = value;
    }

    static get rcInfoUrl(): string {
        return this._rcInfoUrl;
    }

    static set rcInfoUrl(value: string) {
        this._rcInfoUrl = value;
    }

    static get isShowHomeButton(): boolean {
        return this._isShowHomeButton;
    }

    static set isShowHomeButton(value: boolean) {
        this._isShowHomeButton = value;
    }
}
