// general-framework/src/additionalOperatorInformation/AdditionalInfoController.ts
import {EventDispatcher} from "../EventDispatcher";
import {DeviceInfoService} from "../DeviceInfoService";
import {AdditionalInfoCSS} from "./AdditionalInfoCSS";
import {ImageBase64Constants} from "./ImageBase64Constants";
import {LocalUtil} from "../localization/LocalUtil";
import {CommonInitServerResponse} from "../server/controllers/CommonInitServerResponse";

export class AdditionalInfoController {

    public  static readonly AAMS_TICKET_LOCAL_ID: string = "AAMS TICKET ID";
    public  static readonly AAMS_SESSION_LOCAL_ID: string = "AAMS SESSION ID";

    protected static container;
    protected static containerExpanded;

    /**
     * show and create HTML popup
     */
    public static show(): void {
        if ( DeviceInfoService.isMobile
            && CommonInitServerResponse.skinParams
            && (CommonInitServerResponse.skinParams.aams_session || CommonInitServerResponse.skinParams.aams_ticket)
            && ! this.container) {
            this.createPopup();
        }
    }

    /**
     * add listeners
     * 1) response from init end point
     * 2) event about start game
     */
    public static addListener(): void {
        EventDispatcher.getInstance().addListener("RootViewEvents.ADD_ROOT_VIEW_TO_STAGE", this.show.bind(this));
    }

    /**
     * set popup fixed position
     * @param x - right position
     * @param y - top position
     */
    public static setPosition(x:number, y: number): void {
        if (this.containerExpanded ) {
            this.containerExpanded.style.right = x.toString() + "px";
            this.containerExpanded.style.top = y.toString() + "px";
        }
        if (this.container) {
            this.container.style.right = x.toString() + "px";
            this.container.style.top = y.toString() + "px";
        }
    }

    /**
     * set scale for popup
     * @param scaleFactor
     */
    public static setScale(scaleFactor: number): void {
        if (this.containerExpanded ) {
            this.containerExpanded.style.transform = `scale(${scaleFactor},${scaleFactor})`;
        }
        if (this.container) {
            this.container.style.transform = `scale(${scaleFactor},${scaleFactor})`;
        }

    }

    /**
     * create HTML popup
     */
    private static createPopup(): void {
        let style = document.createElement('style');
        style.type = 'text/css';
        style.innerHTML = AdditionalInfoCSS.MAIN;
        document.head.appendChild(style);
        this.containerExpanded = document.createElement("div");
        this.containerExpanded.className = "additionalOperatorInfo";
        this.containerExpanded.innerHTML = "";
        if (CommonInitServerResponse.skinParams.aams_show_ticket) {
            const text: string = CommonInitServerResponse.skinParams.aams_ticket_label || LocalUtil.text(this.AAMS_TICKET_LOCAL_ID);
            this.containerExpanded.innerHTML = this.containerExpanded.innerHTML + `<p class="additionalOperatorInfoParagraph">${text}: ${CommonInitServerResponse.skinParams.aams_ticket}</p>`;
        }
        if (CommonInitServerResponse.skinParams.aams_show_session) {
            const text: string = CommonInitServerResponse.skinParams.aams_session_label || LocalUtil.text(this.AAMS_SESSION_LOCAL_ID);
            this.containerExpanded.innerHTML = this.containerExpanded.innerHTML + `<p class="additionalOperatorInfoParagraph">${text}: ${CommonInitServerResponse.skinParams.aams_session}</p>`;
        }
        this.containerExpanded.innerHTML = this.containerExpanded.innerHTML + `            
            <a href=${CommonInitServerResponse.skinParams.aams_responsible_URL}  target="_blank">
                <img class="operator18Image" id="18_icon" src=${ImageBase64Constants.ICON_18}>
            </a>
            <a href=${CommonInitServerResponse.skinParams.aams_URL} target="_blank">
                <img class="additionalOperatorInfoImage" id="amd_logo" src=${ImageBase64Constants.AMD_LOGO_WHITE}>
            </a>
            
            <img class="additionalOperatorInfoArrows" id="arrow_right" src=${ImageBase64Constants.ARROW_RIGHT}>
            
            <a href=${CommonInitServerResponse.skinParams.aams_URL} target="_blank">
                <img class="baseOperatorInfoImage" id="logo_white" src=${ImageBase64Constants.LOGO_WHITE}>
            </a>
        `;
        document.body.append(this.containerExpanded);
        this.containerExpanded.style.visibility = "hidden";

        this.container = document.createElement("div");
        this.container.className = "additionalOperatorInfo";
        this.container.innerHTML = `        
            
            
            <img class="additionalOperatorInfoArrows" id="arrow_left" src=${ImageBase64Constants.ARROW_LEFT}>
            
            <a href=${CommonInitServerResponse.skinParams.aams_URL} target="_blank">
                <img class="baseOperatorInfoImage" id="logo_white" src=${ImageBase64Constants.LOGO_WHITE}>
            </a>
        `;
        document.body.append(this.container);

        const arrow_right = document.getElementById("arrow_right");
        arrow_right.onclick = (e) => {
            this.containerExpanded.style.visibility = "hidden";
            this.container.style.visibility = "visible";
        }

        const arrow_left = document.getElementById("arrow_left");
        arrow_left.onclick = (e) => {
            this.containerExpanded.style.visibility = "visible";
            this.container.style.visibility = "hidden";
        }

    }

}
