// general-framework/src/additionalOperatorInformation/AdditionalInfoLocalization.ts
/**
 * additional localization
 */
export const AdditionalInfoLocalization = {
    de: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    en: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    es: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    fr: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    gr: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    hi: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    id: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    it: {
        "AAMS TICKET ID": "ID ADM",
        "AAMS SESSION ID": "ID SESSIONE"
    },
    ja: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    ko: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    pl: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    pt: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    ru: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    se: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    sr: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    te: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    th: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    tr: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    vi: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    zh: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    zs: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    },
    hy: {
        "AAMS TICKET ID": "AAMS TICKET ID",
        "AAMS SESSION ID": "AAMS SESSION ID"
    }
}
