// pixi-core-framework/lib/modules/clock/view/ClockView.js
import { BaseView } from "../../../core/engine/BaseView";
import { AdditionalInfoController, DeviceInfoService } from "general-framework";
export class ClockView extends BaseView {
    constructor() {
        super(...arguments);
        this.ADDITIONAL_INFO_OFFSET_X = 20;
        this.ADDITIONAL_INFO_OFFSET_Y = -15;
    }
    initSourceData() {
        this.timeItself = this.source.getChildByName("clockLabel");
        this.timeItselfPortrait = this.source.getChildByName("clockPortraitLabel");
    }
    updateTime() {
        const globalPosition = this.timeItself.getGlobalPosition();
        setTimeout(() => { AdditionalInfoController.setPosition(window.innerWidth - globalPosition.x + this.ADDITIONAL_INFO_OFFSET_X, globalPosition.y + this.ADDITIONAL_INFO_OFFSET_Y); }, 1);
        this.tickUp = null;
        let date = new Date();
        let hours = date.getHours();
        let minutes = date.getMinutes();
        let hoursStr;
        let minutesStr;
        hours < 10 ? hoursStr = "0" + hours : hoursStr = hours.toString();
        minutes < 10 ? minutesStr = "0" + minutes : minutesStr = minutes.toString();
        this.timeItself.setText(hoursStr + ":" + minutesStr);
        if (this.timeItselfPortrait) {
            this.timeItselfPortrait.setText(hoursStr + ":" + minutesStr);
        }
        this.tickUp = setTimeout(this.updateTime.bind(this), 1000);
    }
    onResize(params) {
        if (DeviceInfoService.isMobile && !DeviceInfoService.isLandscape) {
            if (this.timeItself) {
                this.timeItself.visible = false;
            }
            if (this.timeItselfPortrait) {
                this.timeItselfPortrait.visible = true;
            }
        }
        else {
            if (this.timeItself) {
                this.timeItself.visible = true;
            }
            if (this.timeItselfPortrait) {
                this.timeItselfPortrait.visible = false;
            }
        }
    }
}
