// pixi-core-framework/lib/core/baseView/components/pixiProjection/BaseSpriteProjection2d.js
import 'pixi-projection/dist/pixi-projection-spine';
var Sprite2d = PIXI.projection.Sprite2d;
import { TextureHolder } from "../../../utils/layouts/TextureHolder";
import { BaseViewComponentProjection2d } from "./BaseViewComponentProjection2d";
export class BaseSpriteProjection2d extends BaseViewComponentProjection2d {
    constructor() {
        super(...arguments);
        this.initIsComplete = false;
        this.interactiveObject = false;
    }
    init() {
        if (this.initIsComplete) {
            return;
        }
        this._sprite2d = new Sprite2d(TextureHolder.getTextureByName(this._sourceData.image));
        this.addChild(this._sprite2d);
        this.initIsComplete = true;
        if (this._sourceData.width) {
            this.width = this._sourceData.width;
        }
        if (this._sourceData.height) {
            this.height = this._sourceData.height;
        }
        if (this._sourceData.angle !== undefined) {
            this.rotation = this._sourceData.angle * (Math.PI / 180);
        }
        if (this._sourceData.interactive !== undefined) {
            this.interactive = this._sourceData.interactive;
        }
    }
    get sourceData() {
        return this._sourceData;
    }
    get sprite2d() {
        return this._sprite2d;
    }
}
