// pixi-core-framework/lib/core/baseView/components/BaseSprite.js
import { TextureHolder } from "../../utils/layouts/TextureHolder";
import { BaseViewComponent } from "./BaseViewComponent";
/**
 * Framework logic part
 */
export class BaseSprite extends BaseViewComponent {
    constructor() {
        super(...arguments);
        this.initIsComplete = false;
        this.interactiveObject = false;
    }
    init() {
        if (this.initIsComplete) {
            return;
        }
        this.addChild(TextureHolder.getLayoutTextureByName(this._sourceData.image));
        this.initIsComplete = true;
        if (this._sourceData.width) {
            this.width = this._sourceData.width;
        }
        if (this._sourceData.height) {
            this.height = this._sourceData.height;
        }
        if (this._sourceData.angle !== undefined) {
            this.rotation = this._sourceData.angle * (Math.PI / 180);
        }
        if (this._sourceData.interactive !== undefined) {
            this.interactive = this._sourceData.interactive;
        }
    }
    get sourceData() {
        return this._sourceData;
    }
}
