// pixi-core-framework/lib/core/baseView/components/BaseSpriteAnimation.js
import * as PIXI from "pixi.js";
import { BaseViewComponent } from "./BaseViewComponent";
/**
 * Framework logic part
 */
export class BaseSpriteAnimation extends BaseViewComponent {
    constructor(data) {
        super(data);
        this.textures = [];
        this.initIsComplete = false;
        this.frames = data.frames;
        this.framesTotal = data.frames.length;
        if (data.angle !== undefined) {
            this.rotation = data.angle * (Math.PI / 180);
        }
    }
    init() {
        if (this.initIsComplete) {
            return;
        }
        this.frames.forEach((frame) => {
            let texture = PIXI.Texture.fromImage(frame + this.sourceData.extension);
            this.textures.push(texture);
        });
        this.animation = new PIXI.extras.AnimatedSprite(this.textures);
        if (this.sourceData.animationSpeed) {
            this.animation.animationSpeed = this.sourceData.animationSpeed;
        }
        this.addChild(this.animation);
        this.initIsComplete = true;
    }
    set animationSpeed(value) {
        this.animation.animationSpeed = value;
    }
    play(loop = false) {
        return new Promise((resolve) => {
            this.animation.loop = loop;
            this.animation.play();
            if (!loop) {
                this.animation.onComplete = () => {
                    this.animation.gotoAndStop(0);
                    resolve();
                };
            }
        });
    }
    stop() {
        this.animation.stop();
    }
    get sourceData() {
        return this._sourceData;
    }
}
