// pixi-core-framework/lib/core/baseView/components/BaseViewComponent.js
var Container = PIXI.Container;
export class BaseViewComponent extends Container {
    constructor(data) {
        super();
        this._sourceData = data;
        this.pivot.x = data.pivotX;
        this.pivot.y = data.pivotY;
        this.x = data.posX;
        this.y = data.posY;
        this._groupID = data.groupID;
        this._layoutID = data.layoutID;
        if (data.alpha !== undefined) {
            this.alpha = data.alpha;
        }
        if (data.rotation !== undefined) {
            this.rotation = data.rotation;
        }
        if (data.scaleX !== undefined) {
            this.scale.x = data.scaleX;
        }
        if (data.scaleY !== undefined) {
            this.scale.y = data.scaleY;
        }
        if (data.visible !== undefined) {
            this.visible = data.visible;
        }
        if (data.angle !== undefined) {
            this.rotation = data.angle * (Math.PI / 180);
        }
    }
    init() {
    }
    onAdded() {
    }
    get sourceData() {
        return this._sourceData;
    }
    get layoutID() {
        return this._layoutID;
    }
    get groupID() {
        return this._groupID;
    }
    setVisible(value) {
        this.visible = value;
    }
}
