// pixi-core-framework/lib/core/baseView/components/BaseRectangle.js
import * as PIXI from "pixi.js";
import { BaseViewComponent } from "./BaseViewComponent";
/**
 * Framework logic part
 */
export class BaseRectangle extends BaseViewComponent {
    constructor(data) {
        super(data);
        this.initIsComplete = false;
        this.interactiveObject = false;
        if (data.angle !== undefined) {
            this.rotation = data.angle * (Math.PI / 180);
        }
        if (data.interactive !== undefined) {
            this.interactiveObject = data.interactive;
        }
        this.init();
    }
    init() {
        if (this.initIsComplete) {
            return;
        }
        const graphics = new PIXI.Graphics();
        graphics.beginFill(this.sourceData.color || 0xFFFF00, this.sourceData.fillTransparency != undefined ? this.sourceData.fillTransparency : 1);
        if (this.sourceData.lineWidth && this.sourceData.lineWidth > 0) {
            graphics.lineStyle(this.sourceData.lineWidth, this.sourceData.lineColor || 0xFF0000, this.sourceData.lineAlpha || 1);
        }
        if (this.sourceData.cornerRadius) {
            graphics.drawRoundedRect(0, 0, this.sourceData.width, this.sourceData.height, this.sourceData.cornerRadius);
        }
        else {
            graphics.drawRect(0, 0, this.sourceData.width, this.sourceData.height);
        }
        graphics.endFill();
        this.rect = graphics;
        this.addChild(graphics);
        this.initIsComplete = true;
        this.rect.interactive = this.interactiveObject;
    }
    get sourceData() {
        return this._sourceData;
    }
    get rect() {
        return this._rect;
    }
    set rect(value) {
        this._rect = value;
    }
}
