// pixi-core-framework/lib/core/baseView/components/RangeComponent.js
var Container = PIXI.Container;
import { PointerEventsConstants } from "../PointerEventsConstants";
import { RangeComponentConstants } from "../RangeComponentConstants";
export class RangeComponent extends Container {
    constructor(source) {
        super();
        this.source = source;
        this.increaseButton = source.getChildByName(RangeComponentConstants.INCREASE_BUTTON);
        this.increaseButton.on(PointerEventsConstants.CLICKED, (value) => {
            this.emit(RangeComponent.INCREASE_CLICKED);
        });
        this.decreaseButton = source.getChildByName(RangeComponentConstants.DECREASE_BUTTON);
        this.decreaseButton.on(PointerEventsConstants.CLICKED, (value) => {
            this.emit(RangeComponent.DECREASE_CLICKED);
        });
        this.valueLabel = source.getChildByName(RangeComponentConstants.VALUE_LABEL);
        this.progressBar = source.getChildByName(RangeComponentConstants.PROGRESS_BAR);
    }
    updateComponent(data) {
        if (data.decrease !== undefined) {
            this.decreaseButton.enable(data.decrease);
            if (data.decrease) {
                this.decreaseButton.setVisible(true);
            }
        }
        if (data.increase !== undefined) {
            this.increaseButton.enable(data.increase);
            if (data.increase) {
                this.increaseButton.setVisible(true);
            }
            ;
        }
        data.labelValue !== undefined && this.valueLabel.setText(data.labelValue);
        this.progressBar.rect.scale.x = data.progressPercent;
    }
    setVisible(value) {
        this.source.visible = value;
    }
    enable(value) {
        if (value) {
            this.decreaseButton.enable(this.decreaseButtonEnable);
            this.increaseButton.enable(this.increaseButtonEnable);
        }
        else {
            this.decreaseButtonEnable = this.decreaseButton.isEnabled;
            this.increaseButtonEnable = this.increaseButton.isEnabled;
            this.decreaseButton.enable(false);
            this.increaseButton.enable(false);
        }
    }
}
RangeComponent.INCREASE_CLICKED = "INCREASE_CLICKED";
RangeComponent.DECREASE_CLICKED = "DECREASE_CLICKED";
