// pixi-core-framework/lib/core/utils/dateUtil/DateUtil.js
export class DateUtil {
    static parseServerDate(data) {
        return this.getDate(data) + "\n" + this.getTime(data);
    }
    static getDate(data) {
        let date = new Date(data);
        let year = date.getFullYear().toString();
        let month = this.toDoubleDigitFormat(date.getMonth() + 1);
        let day = this.toDoubleDigitFormat(date.getDate());
        return year + "-" + month + "-" + day;
    }
    static getTime(data) {
        let date = new Date(data);
        let hours = this.toDoubleDigitFormat(date.getHours());
        let minutes = this.toDoubleDigitFormat(date.getMinutes());
        let seconds = this.toDoubleDigitFormat(date.getSeconds());
        return hours + ":" + minutes + ":" + seconds;
    }
    static toDoubleDigitFormat(value) {
        return value >= 10 ? value.toString() : "0" + value;
    }
    static getTimezoneOffset(data) {
        let sign = data.substr(19, 1);
        let offset = parseInt(data.substr(20, 2));
        return sign === "+" ? offset : offset * -1;
    }
}
