// general-framework/src/logger/AppLogger.ts
import { AppLoggerMessageType } from './AppLoggerMessageType';

const Colors = {
    ERROR: "background: red;color: white",
    EVENT: "background: blue;color: white",
    SUCCESS: "background: green; color: white",
    DEFAULT: ""
}

export class AppLogger {

    /**
     * Logs application message
     * @param message actual message to log
     * @param [type] optional type
     * @param [obj] optional data
     */
    public static log(message: string, type?: string, params?: any): void {
        if (process.env.NODE_ENV === 'production') {
            return;
        }
        type = AppLoggerMessageType[type] ? type : AppLoggerMessageType.DEFAULT;
        let stringParams: string = "";
        if (params) {
            try {
                stringParams = JSON.stringify(params);
            }
            catch(e)
            {
                stringParams = "can't stringify params";
            }
        }
        console.log(`%c${message}${stringParams}`, Colors[type]);
    }

}
