// pixi-core-framework/lib/modules/clock/events/ClockEvents.js
export const ClockEvents = {
    /**
     * @START_CLOCK
     *
     * Clock start update / initializing event
     * Should be sent with some of options {
     *      options{
     *      weekday:{"long","short","narrow"},
     *      year:{"numeric","2-digit"},
     *      month:{"numeric","2-digit"},
     *      day:{"numeric","2-digit"},
     *      second: {"numeric","2-digit"},
     *      hour: {"numeric","2-digit"}
     *      minute: {"numeric","2-digit"},
     *      timeZone: ""}
     * }
     */
    START_CLOCK: "ClockEvents.START_CLOCK",
    /**
     * @STOP_CLOCK
     * Clock stop update
     */
    STOP_CLOCK: "ClockEvents.STOP_CLOCK",
    /**
     * @SHOW_TIME
     * Clock view update event
     */
    SHOW_TIME: "ClockEvents.SHOW_TIME",
    UPDATE_TIME: "ClockEvents.UPDATE_TIME",
    /**
     * @ALERT_EVENT
     *
     * This event is used to start any custom timer to dispatch incoming event.
     * Should be sent with {
     *      @interval - time after timer stop in seconds
     *      @event - event name which dispatched after timer stop
     *      @repeats - number of cycles to repeat same timer
     *      @loop - loop timer till application is open, if timer is loop, repeats are ignored
     * }
     *
     */
    ALERT_EVENT: "ClockEvents.ALERT_EVENT"
};
