// general-framework/lib/reality-check/RealityCheckModel.js
/**
 * Save and return values for reality checking
 */
export class RealityCheckModel {
    static get time_elapsed() {
        return this._time_elapsed ? this._time_elapsed : 0;
    }
    static set time_elapsed(value) {
        this._time_elapsed = value;
    }
    static get time_limit() {
        return this._time_limit ? this._time_limit : null;
    }
    static set time_limit(value) {
        this._time_limit = value;
    }
    static get exit_URL() {
        return this._exit_URL;
    }
    static set exit_URL(value) {
        this._exit_URL = value;
    }
    static get lang() {
        return this._lang;
    }
    static set lang(value) {
        this._lang = value;
    }
    static get isFrozen() {
        return this._isFrozen;
    }
    static set isFrozen(value) {
        this._isFrozen = value;
    }
    static get startTime() {
        return this._startTime;
    }
    static set startTime(value) {
        this._startTime = value;
    }
    static get lobby_URL() {
        return this._lobby_URL;
    }
    static set lobby_URL(value) {
        this._lobby_URL = value;
    }
    static get history_URL() {
        return this._history_URL;
    }
    static set history_URL(value) {
        this._history_URL = value;
    }
    static get rcInfoUrl() {
        return this._rcInfoUrl;
    }
    static set rcInfoUrl(value) {
        this._rcInfoUrl = value;
    }
    static get isShowHomeButton() {
        return this._isShowHomeButton;
    }
    static set isShowHomeButton(value) {
        this._isShowHomeButton = value;
    }
}
