// general-framework/lib/additionalOperatorInformation/AdditionalInfoCSS.js
/**
 * embedded css
 */
export const AdditionalInfoCSS = {
    MAIN: `.additionalOperatorInfo {
            display: flex;
            justify-content: flex-end;
            align-items: center;
            position: fixed;
            outline: none;            
            margin: 0;
            transform-origin: center right;
            background-color: rgba(28, 28, 28, 0.6);
            z-index: 2;
            
            border-radius: 43px;
            border: 2px solid rgba(92, 92, 92, 0.6);
            height: 52px;
            
            top: 15px;
            right: 75px;
            }
            
            .additionalOperatorInfoParagraph {
                margin-left: 24px;
                margin-right: 13px;
                margin-bottom: 9px;
                margin-top: 5px;
                font-family: Open Sans;
                font-style: normal;
                font-weight: bold;
                font-size: 10px;
                line-height: 14px;
                height: 14px;
                width: max-content;
            }
            .baseOperatorInfoImage {
                margin-left: 19px;
                margin-right: 21px;
                margin-bottom: 1px;
                margin-top: 8px;
                height: max-content;
            }
            .additionalOperatorInfoImage {
                margin-left: 19px;
                margin-right: 13px;
                margin-bottom: 1px;
                margin-top: 3px;
                height: max-content;
            }            
            .operator18Image {
                margin-left: 19px;
                margin-right: 13px;
                margin-bottom: 1px;
                margin-top: 7px;
                height: max-content;
            }
            .additionalOperatorInfoArrows {
                margin-left: 19px;
                margin-right: 24px;
                margin-bottom: 3x;
                margin-top: 1px;
                height: max-content;
            }
            `
};
