// general-framework/webpack/build.conf.js
const path = require('path');

const merge = require('webpack-merge');
const CompressionPlugin = require('compression-webpack-plugin');
const CopyPlugin = require('copy-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const {CleanWebpackPlugin} = require('clean-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const OptimizeCSSAssetsPlugin = require('optimize-css-assets-webpack-plugin');
const TerserPlugin = require('terser-webpack-plugin');

const LinkPreparePlugin = require('./plugins/rel-prepare');
const MakeInfoFilePlugin = require('./plugins/makeInfoFile');
// const imgMinify = require('./plugins/img-minify');

const baseConfig = require('./base.conf');
const PATHS = require("./constants");

const packageJsonPath = path.resolve(PATHS.root, "package.json");
const {dependencies, devDependencies} = require(packageJsonPath);
const packages = {...dependencies, ...devDependencies};
const outputDir = process.env.NODE_ENV === 'production' ? PATHS.buildProd : PATHS.build;
const buildTimestamp = Date.now();

const buildWebpackConfig = merge(baseConfig, {
    output: {
        filename: 'js/[name].[hash].js',
        path: outputDir,
    },
    plugins: [
        new CopyPlugin([
            {
                from: PATHS.data,
                to: 'data',
                transformPath(targetPath, absolutePath) {
                    return PATHS.cacheList.some((part) => targetPath.includes(part))
                        ? path.join('data', buildTimestamp.toString(), targetPath.replace('data', ''))
                        : targetPath;
                },
                // to: 'data/[name].[hash].[ext]',
                // transform(content, srcPath) {
                //     return process.env.NODE_ENV === 'production' && /\.(png|jpe?g)$/.test(path.extname(srcPath))
                //         ? imgMinify(content)
                //         : content
                // },
            },
        ]),
        new MiniCssExtractPlugin({
            filename: '[name].[hash].css',
            chunkFilename: '[id].[hash].css',
        }),
        new LinkPreparePlugin({
            root: PATHS.root,
            output: outputDir,
            buildTimestamp
        }),
        new MakeInfoFilePlugin({
            output: outputDir,
            buildTimestamp
        }),
        new HtmlWebpackPlugin({
            favicon: baseConfig.externals.favicon,
            template: PATHS.src + '/index.html',
            filename: 'index.html',
            minify: {
                collapseWhitespace: true,
                removeComments: true,
                minifyCSS: true,
            },
            inject: true,
        }),
        process.env.NODE_ENV === 'production' && new CompressionPlugin({
            deleteOriginalAssets: false,
            filename: (info) => `${info.file}.gz`,
            // TODO: remove html after rework link-rel-maker
            exclude: /\.(html)$/
        }),
        new CleanWebpackPlugin(),
    ].filter(Boolean),
    optimization: {
        minimize: true,
        minimizer: [
            new TerserPlugin({
                sourceMap: process.env.NODE_ENV === 'development',
                parallel: true,
                terserOptions: {
                    parse: {
                        ecma: 8,
                    },
                    compress: {
                        ecma: 5,
                        warnings: false,
                        comparisons: false,
                        inline: 2,
                        drop_console: process.env.NODE_ENV === 'production',
                    },
                    mangle: {
                        safari10: true,
                    },
                    output: {
                        ecma: 5,
                        comments: false,
                    },
                },
            }),
            new OptimizeCSSAssetsPlugin(),
        ],
        // splitChunks: {
        //     cacheGroups: {
        //         vendor: {
        //             name: 'vendors',
        //             test: /node_modules/,
        //             chunks: 'all',
        //             enforce: true
        //         },
        //         // styles: {
        //         //     name: 'styles',
        //         //     test: /\.css$/,
        //         //     chunks: 'all',
        //         //     enforce: true,
        //         // },
        //     },
        // },
        /* use for multiple chunks */
        runtimeChunk: 'single',
        splitChunks: {
            chunks: 'all',
            maxInitialRequests: 6,
            minSize: 30000,
            cacheGroups: {
                vendor: {
                    test: /[\\/]node_modules[\\/]/,
                    name(module, chunks, cacheGroupKey) {
                        const modulePath = module.context.match(/[\\/]node_modules[\\/](.*?)([\\/]|$)/)[1];
                        let moduleName = modulePath.replace("@", "");
                        let result;

                        if (packages.hasOwnProperty(moduleName)) {
                            const version = packages[moduleName]

                            if (version.match('#')) {
                                result = version.split("#")[1];
                            } else if (version.toLowerCase().match(/[a-z]/g)) {
                                result = 'latest'
                            } else {
                                result = version;
                            }

                            result = result.replace(/[*^]/g, '')
                        }

                        return `npm.${moduleName}${result ? '.v:' + result : ''}`;
                    }
                },
            },
        },
    },
    devtool: 'source-map',
});

module.exports = Promise.resolve(buildWebpackConfig);
