// general-framework/src/loading/controllers/HtmlPreloaderController.ts
import {LoaderConstants} from "../constants/LoaderConstants";
import {LoadEvents} from '../events/LoadEvents';
import {EventDispatcher} from '../../EventDispatcher';

export class HtmlPreloaderController {
    current: number;
    gameLogo: HTMLElement;
    loaderElement: HTMLElement;
    videoContainer: Element;
    protected isPaused: boolean = false;
    protected canBeRemoved: boolean = false;
    private dispatcher = EventDispatcher.getInstance();

    constructor() {
        this.current = 1;
        this.gameLogo = document.getElementById('gameLogo');
        this.loaderElement = document.getElementById('logo-progress-bar');
        this.videoContainer = this.gameLogo.getElementsByClassName('video-container')[0];
        const logoVideo: HTMLMediaElement = document.querySelector('#gameLogo video');

        if (window["logoAnimationIsPlayed"]) {
            logoVideo.addEventListener("pause", () => {
                this.isPaused = true;
                if (this.canBeRemoved) {
                    this.startFade();
                }
            })
        } else {
            this.isPaused = true;
        }
        this.addListeners();
    }

    protected addListeners(): void {
        this.dispatcher.addListener(LoadEvents.LOAD_PROGRESS, this.updateProgress, this);
        this.dispatcher.addListener(LoadEvents.REMOVE_PRELOADER, this.hideLoader, this);
    }

    protected updateProgress(progress: number): void {
        let currentProgress: number = this.current / progress * 100;
        this.loaderElement.style.width = currentProgress + '%';
        this.dispatcher.dispatch(LoadEvents.UPDATE_LOAD_PROGRESS, currentProgress);
        this.current++;
    }

    protected hideLoader(priority: string): void {

        if (priority === LoaderConstants.PRELOAD_PRIORITY) {

            this.dispatcher.removeListener(LoadEvents.LOAD_PROGRESS, this.updateProgress, this);
            this.canBeRemoved = true;
            if (!this.isPaused) {
                return
            }
            this.startFade();
        }
    }

    protected startFade(): void {
        setTimeout(() => {
            this.videoContainer.classList.add('hide');
            setTimeout(() => {
                this.gameLogo.classList.add('hide');
                setTimeout(() => {
                    this.gameLogo.remove();
                }, 800);
            }, 800);
        }, 300);
    }
}
