// general-framework/src/additionalOperatorInformation/models/FrontSettingsModel.ts
export class FrontSettingsModel {

    private static _isShowBalanceMobile: boolean;
    private static _isShowBalanceDesktop: boolean;

    private static _isHiddenCompletely: boolean;

    private static _isCountTimeInAutoSpinMode: boolean;


    public static parseInitResponse(data): void {
        if (data.isShowBalance) {
            this._isShowBalanceDesktop = data.isShowBalance.isShowBalanceDesktop;
            this._isShowBalanceMobile = data.isShowBalance.isShowBalanceMobile;

            this._isHiddenCompletely = data.isShowBalance.isHiddenCompletely;
        }

        if (data.isCountTimeInAutoSpinMode) {
            this._isCountTimeInAutoSpinMode = data.isCountTimeInAutoSpinMode;
        }

    }

    static get isHiddenCompletely(): boolean {
        return this._isHiddenCompletely;
    }

    static get isShowBalanceDesktop(): boolean {
        return this._isShowBalanceDesktop;
    }

    static get isShowBalanceMobile(): boolean {
        return this._isShowBalanceMobile;
    }

    static get isCountTimeInAutoSpinMode(): boolean {
        return this._isCountTimeInAutoSpinMode;
    }
}
