// pixi-core-framework/src/modules/clock/view/ClockView.ts
import {BaseView} from "../../../core/engine/BaseView";
import {BaseLabel} from "../../../core/baseView/components/BaseLabel";
import Point = PIXI.Point;
import {AdditionalInfoController, DeviceInfoService} from "general-framework";
import {IRenderViewParams} from "../../renderModule/IRenderViewParams";

export class ClockView extends BaseView {
    protected timeItself: BaseLabel;
    protected timeItselfPortrait: BaseLabel;
    protected tickUp;
    protected ADDITIONAL_INFO_OFFSET_X: number = 20;
    protected ADDITIONAL_INFO_OFFSET_Y: number = -15;

    protected initSourceData(): void {
        this.timeItself = this.source.getChildByName("clockLabel");
        this.timeItselfPortrait = this.source.getChildByName("clockPortraitLabel");
    }

    public updateTime() {

        const globalPosition: Point = this.timeItself.getGlobalPosition();
        setTimeout(()=>{AdditionalInfoController.setPosition(window.innerWidth - globalPosition.x + this.ADDITIONAL_INFO_OFFSET_X, globalPosition.y + this.ADDITIONAL_INFO_OFFSET_Y)}, 1 );

        this.tickUp = null;

        let date = new Date();

        let hours: Number = date.getHours();
        let minutes: Number = date.getMinutes();

        let hoursStr: string;
        let minutesStr: string;

        hours < 10 ? hoursStr = "0" + hours : hoursStr = hours.toString();
        minutes < 10 ? minutesStr = "0" + minutes : minutesStr = minutes.toString();

        this.timeItself.setText(hoursStr + ":" + minutesStr);
        if (this.timeItselfPortrait) {
            this.timeItselfPortrait.setText(hoursStr + ":" + minutesStr);
        }
        this.tickUp = setTimeout(this.updateTime.bind(this), 1000);
    }

    public onResize(params: IRenderViewParams) {
        if (DeviceInfoService.isMobile && !DeviceInfoService.isLandscape) {
            if (this.timeItself) {
                this.timeItself.visible = false;
            }
            if (this.timeItselfPortrait) {
                this.timeItselfPortrait.visible = true;
            }
        } else {
            if (this.timeItself) {
                this.timeItself.visible = true;
            }
            if (this.timeItselfPortrait) {
                this.timeItselfPortrait.visible = false;
            }
        }
    }
}
