// pixi-core-framework/src/core/baseView/components/BaseRectangle.ts
import {IRectangleConfig} from "../interfaces/IRectangleConfig";
import Graphics = PIXI.Graphics;
import * as PIXI from "pixi.js";
import {BaseViewComponent} from "./BaseViewComponent";

/**
 * Framework logic part
 */
export class BaseRectangle extends BaseViewComponent {
    protected _sourceData: IRectangleConfig;
    private _rect: Graphics;
    protected initIsComplete: boolean = false;
    protected interactiveObject: boolean = false;

    constructor(data: IRectangleConfig) {
        super(data);
        if (data.angle !== undefined) {
            this.rotation = data.angle * (Math.PI / 180);
        }
        if (data.interactive !== undefined) {
            this.interactiveObject = data.interactive;
        }
        this.init();
    }

    public init(): void {
        if (this.initIsComplete) {
            return;
        }
        const graphics = new PIXI.Graphics();
        graphics.beginFill(this.sourceData.color || 0xFFFF00, this.sourceData.fillTransparency != undefined ? this.sourceData.fillTransparency : 1);
        if (this.sourceData.lineWidth && this.sourceData.lineWidth > 0) {
            graphics.lineStyle(this.sourceData.lineWidth, this.sourceData.lineColor || 0xFF0000, this.sourceData.lineAlpha || 1);
        }
        if (this.sourceData.cornerRadius) {
            graphics.drawRoundedRect(0, 0, this.sourceData.width, this.sourceData.height, this.sourceData.cornerRadius);
        } else {
            graphics.drawRect(0, 0, this.sourceData.width, this.sourceData.height);
        }
        graphics.endFill();
        this.rect = graphics;
        this.addChild(graphics);
        this.initIsComplete = true;
        this.rect.interactive = this.interactiveObject;
    }

    get sourceData(): IRectangleConfig {
        return this._sourceData;
    }

    get rect(): PIXI.Graphics {
        return this._rect;
    }

    set rect(value: PIXI.Graphics) {
        this._rect = value;
    }
}
