// pixi-core-framework/lib/modules/backgrounds/BackgroundsView.js
import { BaseView } from "../../core/engine/BaseView";
import { CoreRenderer } from "../../core/engine/renderer/CoreRenderer";
import { TweenLite } from "gsap";
import { getViewComponent } from "../../core/engine/createInstance";
/**
 * Framework logic part
 */
export class BackgroundsView extends BaseView {
    constructor() {
        super(...arguments);
        this.animationTween = [];
    }
    initSourceData() {
        this.currentBackground = this.source.getChildByName("main_background");
        CoreRenderer.stage.addChild(this.source);
        this.hide();
    }
    onResize(params) {
        this.currentResizeParams = params;
        if (params.width && params.height) {
            this.source.scale.x = this.source.scale.y = params.scale;
        }
        let currentWidth = document.documentElement.clientWidth;
        let currentHeight = document.documentElement.clientHeight;
        const currentRatio = currentHeight / currentWidth;
        const defaultRatio = this.defaultHeight / this.defaultWidth;
        if (currentWidth > this.defaultWidth * params.scale || currentHeight > this.defaultHeight * params.scale) {
            if (defaultRatio > currentRatio) {
                this.source.width = currentWidth;
                this.source.height = Math.round(this.source.width * this.defaultHeight / this.defaultWidth);
            }
            else {
                this.source.height = currentHeight;
                this.source.width = Math.round(this.source.height * this.defaultWidth / this.defaultHeight);
            }
        }
        this.source.position.x = currentWidth / 2 - this.source.width / 2;
        this.source.position.y = currentHeight / 2 - this.source.height / 2;
    }
    changeBackground(backgroundID) {
        this.backgroundToShow = this.source.getChildByName(backgroundID + "_background");
        if (this.backgroundToShow.children.length === 0) {
            let viewComponent = getViewComponent(backgroundID + "_background_content");
            viewComponent.children.forEach((item) => {
                item.init();
            });
            viewComponent.init();
            viewComponent.vissible = true;
            this.backgroundToShow.addChild(viewComponent);
        }
        this.defaultWidth = Math.round(this.backgroundToShow.width);
        this.defaultHeight = Math.round(this.backgroundToShow.height);
        this.backgroundToShow.alpha = 0;
        this.source.addChild(this.backgroundToShow);
        this.startAnimation(this.backgroundToShow);
    }
    startAnimation(backgroundToShow) {
        this.animationTween.push(TweenLite.to(backgroundToShow, .6, {
            alpha: 1, onComplete: () => {
            }
        }));
        this.onResize(this.currentResizeParams);
    }
}
