// general-framework/lib/CheckFewTabs.js
import { GameInfo } from "./GameInfo";
import { EventDispatcher } from "./EventDispatcher";
import { CoreEvents } from "./CoreEvents";
/**
 * Checking on game opened in more than one tab.
 * Need to create it instance after GameInfo class is initialized
 * Automatically checked on change storage event (not working in current tab) and on GameInfo was initialized
 */
export class CheckFewTabs {
    constructor() {
        this.dispatcher = EventDispatcher.getInstance();
        this.prevTimestamp = Date.now();
        this.isTwoTabs = false;
        this.checkOnFewTabs = (event) => {
            if (GameInfo.title && event.key === GameInfo.title) {
                const { token, timestamp } = JSON.parse(event.newValue);
                if (token === GameInfo.token && this.prevTimestamp < timestamp) {
                    this.isTwoTabs = true;
                    this.dispatcher.dispatch(CoreEvents.GAME_IN_FEW_TABS);
                }
            }
        };
        try {
            localStorage.setItem(GameInfo.title, JSON.stringify({
                token: GameInfo.token,
                timestamp: this.prevTimestamp
            }));
        }
        catch (err) {
            console.log(err);
        }
        window.addEventListener("storage", this.checkOnFewTabs);
        this.dispatcher.addListener(CoreEvents.GAME_INFO_INITIALIZED, this.checkOnFewTabs);
    }
}
