// general-framework/lib/server/controllers/CommonInitServerResponse.js
import { RealityCheckModel } from "../../reality-check/RealityCheckModel";
import { FrontSettingsModel } from "../../additionalOperatorInformation/models/FrontSettingsModel";
import { PingController } from "../../operator/PingController";
import { RealityCheckData } from "../../reality-check/RealityCheckData";
import { APIController } from "../../operator/APIController";
export class CommonInitServerResponse {
    static parseResponse(data) {
        const initData = data.data;
        if (initData) {
            if (initData.gameParams && initData.gameParams["front.settings"]) {
                this.parseGameSettings(initData.gameParams["front.settings"]);
            }
            if (initData.gameParams && initData.gameParams["skin_params"]) {
                this.parseOperatorSettings(initData.gameParams["skin_params"]);
            }
            if (initData.gameParams && initData.gameParams.APIData) {
                APIController.APIData = initData.gameParams.APIData;
            }
        }
    }
    /**
     * Method parse Reality check ond other "front.settings" params from response
     * @param data - server init response
     */
    static parseGameSettings(data) {
        if (data.realityCheck) {
            this._realityCheckData = data.realityCheck;
            this.setRealityCheckData(this._realityCheckData);
        }
        FrontSettingsModel.parseInitResponse(data);
    }
    /**
     * Method parse Reality check ond other skin params from response
     * @param data - server init response
     */
    static parseOperatorSettings(data) {
        let realityCheckData = data;
        this._skinParams = data;
        if (realityCheckData.time_limit !== undefined) {
            this._realityCheckData = realityCheckData;
            this.setRealityCheckData(this._realityCheckData);
        }
        if (this._skinParams) {
            if (this._skinParams.operator_ping) {
                PingController.data = typeof this._skinParams.operator_ping === 'string'
                    ? JSON.parse(this._skinParams.operator_ping)
                    : this._skinParams.operator_ping;
            }
            if (this._skinParams.game_provider_id) {
                RealityCheckData.game_provider_id = this._skinParams.game_provider_id;
            }
        }
    }
    /**
     * Method set params to model
     * @param data - reality check params
     */
    static setRealityCheckData(data) {
        RealityCheckModel.exit_URL = data.exit_URL ? decodeURIComponent(data.exit_URL) : null;
        RealityCheckModel.lobby_URL = data.lobby_URL ? decodeURIComponent(data.lobby_URL) : null;
        RealityCheckModel.history_URL = data.history_URL ? decodeURIComponent(data.history_URL) : null;
    }
    static get skinParams() {
        return this._skinParams;
    }
}
