// general-framework/lib/loading/events/LoadEvents.js
export const LoadEvents = {
    INIT_LOADER: "LoadEvents.INIT_LOADER",
    LOAD_SPINE_DATA: "LoadEvents.LOAD_SPINE_DATA",
    SPINE_DATA_LOADED: "LoadEvents.SPINE_DATA_LOADED",
    LOAD_PROGRESS: "LoadEvents.LOAD_PROGRESS",
    UPDATE_LOAD_PROGRESS: "LoadEvents.UPDATE_LOAD_PROGRESS",
    LOAD_CONFIGS: "LoadEvents.LOAD_CONFIGS",
    CONFIG_LOADED: "LoadEvents.CONFIG_LOADED",
    LOAD_SOUNDS: "LoadEvents.LOAD_SOUNDS",
    SOUND_LOADED: "LoadEvents.SOUND_LOADED",
    LOAD_FONTS: "LoadEvents.LOAD_FONTS",
    FONTS_LOADED: "LoadEvents.FONTS_LOADED",
    LOAD_ASSETS: "LoadEvents.LOAD_ASSETS",
    LOAD_BITMAP_FONTS: "LoadEvents.LOAD_BITMAP_FONTS",
    ASSET_LOADED: "LoadEvents.ASSET_LOADED",
    REMOVE_PRELOADER: "LoadEvents.REMOVE_PRELOADER",
    BITMAP_FONTS_LOADED: "LoadEvents.BITMAP_FONTS_LOADED",
    FORCE_LOAD: "LoadEvents.FORCE_LOAD",
    FORCE_LOAD_COMPLETE: "LoadEvents.FORCE_LOAD_COMPLETE",
    LOAD_LOCALIZATION: "LoadEvents.LOAD_LOCALIZATION",
    LOCALIZATION_LOADED: "LoadEvents.LOCALIZATION_LOADED",
    LOADING_COMPLETE: "LoadEvents.LOADING_COMPLETE",
    INITIAL_LOADING_COMPLETE: "LoadEvents.INITIAL_LOADING_COMPLETE",
    INIT_MOBILE_MANAGER: "LoadEvents.INIT_MOBILE_MANAGER",
    RESET_LOADER_VIEW: "LoadEvents.RESET_LOADER_VIEW",
    UPDATE_POSITION: "LoadEvents.UPDATE_POSITION",
    HIDE_LOADER: "LoadEvents.HIDE_LOADER",
    GAME_PARAMS_DATA_RECEIVED: "LoadEvents.GAME_PARAMS_DATA_RECEIVED",
    SPINE_FACTORY_DATA_RECEIVED: "LoadEvents.SPINE_FACTORY_DATA_RECEIVED",
    VIEW_COMPONENTS_FACTORY_DATA_RECEIVED: "LoadEvents.VIEW_COMPONENTS_FACTORY_DATA_RECEIVED",
    ATLAS_LOADING_STARTED: "LoadEvents.ATLAS_LOADING_STARTED",
    PROJECT_LOADING_COMPLETE: "LoadEvents.PROJECT_LOADING_COMPLETE"
};
