// pixi-core-framework/lib/core/utils/screenMobileManager/ScreenMobileManager.js
import { BaseController } from "../../engine/BaseController";
import { DeviceInfoService, LoadEvents } from "general-framework";
import { AppLogger } from "../logger/AppLogger";
import { AppLoggerMessageType } from "../logger/AppLoggerMessageType";
export class ScreenMobileManager extends BaseController {
    constructor() {
        super();
        this.screen = window.screen;
        this.canvas = document.getElementById('canvas-wrap');
        this.btnFullScreen = document.getElementById('full-screen-btn');
        this.rotateScreen = document.getElementById('phoneOverlay');
        this.dispatcher.addListener(LoadEvents.INIT_MOBILE_MANAGER, () => {
            if (DeviceInfoService.isMobile) {
                // need to avoid handling the iOS Chrome - it not correct counts the innerWidth/Height on resize
                if (DeviceInfoService.isIOs && DeviceInfoService.isChrome)
                    return;
                if (DeviceInfoService.isIOs) {
                    this.checkIframe();
                    if (!DeviceInfoService.isLandscape || DeviceInfoService.isLandscapeFullScreen) {
                        this.hideBtn();
                    }
                    else {
                        this.showBtn();
                        this.addScrollUpListener();
                    }
                }
                else {
                    this.btnFullScreen.addEventListener("touchend", () => {
                        this.hideBtn();
                        ScreenMobileManager.turnOnFullScreen();
                    });
                    this.btnFullScreen.addEventListener("click", () => {
                        this.hideBtn();
                        ScreenMobileManager.turnOnFullScreen();
                    });
                }
                if (this.screen.orientation) {
                    // only for chrome, firefox
                    // automatically enable full screen
                    window.addEventListener("orientationchange", this.onOrientationChange.bind(this));
                }
                else {
                    // Important: on mobile resize event can occurs several times on change orientation
                    window.addEventListener("resize", this.onResize.bind(this));
                }
                this.handleFullScreenBtn();
                this.handleFullScreenOut();
            }
        });
    }
    addScrollUpListener() {
        let isStarted = false;
        this.btnFullScreen.addEventListener("touchstart", () => {
            isStarted = true;
        });
        this.btnFullScreen.addEventListener("touchend", () => {
            isStarted = false;
        });
        window.addEventListener("resize", () => {
            this.hideBtn();
        });
    }
    checkIframe() {
        if (window.location !== window.parent.location && DeviceInfoService.isMobile) {
            this.hideBtn();
        }
    }
    onResize() {
        this.handleFullScreenBtn();
        if (DeviceInfoService.isIOs) {
            this.checkIframe();
        }
    }
    ;
    onOrientationChange() {
        if (!DeviceInfoService.isIOs && !document.fullscreenElement) {
            this.showBtn();
            setTimeout(() => {
                this.checkCanRotate();
            }, 500);
        }
    }
    ;
    handleFullScreenOut() {
        document.addEventListener('fullscreenchange', (event) => {
            if (!DeviceInfoService.isIOs && !document.fullscreenElement) {
                this.showBtn();
            }
        });
    }
    /**
     * safari mobile/tablet dimensions:
     [320, 568],
     [375, 667],
     [375, 812],
     [414, 736],
     [768, 1024],
     [1024, 1366],
     [1112, 834]
     */
    // private handleIOsFullScreen() {
    //     const ratioScreen = (screen.width / screen.height).toFixed(2); // screen values aren't flipping by change orientation
    //     const {innerHeight: h, innerWidth: w} = window;
    //     const ratioInner = (w > h ? h / w : w / h).toFixed(2);
    //     (ratioInner === ratioScreen) ? this.hideBtn() : this.showBtn();
    // }
    handleFullScreenBtn() {
        if (!DeviceInfoService.isIOs) {
            if (!document.fullscreenElement) {
                this.showBtn();
            }
            /*if (DeviceInfoService.isLandscape) {

            }*/
        }
    }
    showBtn() {
        this.canvas.style.pointerEvents = 'none';
        this.btnFullScreen.style.display = 'block';
    }
    hideBtn() {
        this.canvas.style.pointerEvents = 'auto';
        this.btnFullScreen.style.display = 'none';
    }
    // attention: fullScreen API not supported for iOS Safari
    // it partition supported only for iPad
    static turnOnFullScreen() {
        const elem = document.documentElement;
        if (elem.requestFullscreen) {
            elem.requestFullscreen().catch(err => {
                AppLogger.log("requestFullscreen doesn't work!", AppLoggerMessageType.ERROR);
            });
        }
        else if (elem.mozRequestFullScreen) {
            elem.mozRequestFullScreen();
        }
        else if (elem.webkitRequestFullscreen) {
            elem.webkitRequestFullscreen();
        }
        else if (elem.msRequestFullscreen) {
            elem.msRequestFullscreen();
        }
        this.isOnFullScreenmode = true;
    }
    static turnOffFullScreen() {
        if (document.exitFullscreen) {
            document.exitFullscreen();
            // @ts-ignore
        }
        else if (document.mozCancelFullScreen) {
            // @ts-ignore
            document.mozCancelFullScreen();
        }
        else if (document.webkitExitFullscreen) {
            document.webkitExitFullscreen();
            // @ts-ignore
        }
        else if (document.msExitFullscreen) {
            // @ts-ignore
            document.msExitFullscreen();
        }
        this.isOnFullScreenmode = false;
    }
    checkCanRotate() {
        if (!DeviceInfoService.isLandscape && this.rotateScreen.style.visibility !== "hidden") {
            this.hideBtn();
        }
    }
}
