// pixi-core-framework/src/core/engine/fsm/SerialAction.ts
import {BaseAction} from "./BaseAction";

/**
 * This action is resolved after the last included action is resolved.
 * The actions start step by step in a queue order.
 */
export class SerialAction extends BaseAction {

    protected executedQueue: BaseAction[] = [];
    protected mainResolve: Function;
    protected currentAction: BaseAction;

    execute(): Promise<any> {
        return new Promise<void>(resolve => {
            this.queue().then(() => {
                this.end();
                resolve();
            })
        });
    }

    queue(): Promise<void> {
        return new Promise<void>( resolve => {
            this.executedQueue = this.actions.concat();
            this.executeActionFromQueue();
            this.mainResolve = resolve;
        });
    }

    public skip(): void {
        super.skip();
        this.currentAction.skip();
    }

    protected executeActionFromQueue(): void {
        if (this.executedQueue.length === 0 || this.shouldSkip) {
            this.mainResolve();
            return;
        }
        this.currentAction = this.executedQueue.shift();
        this.currentAction.execute().then(() => {
            this.executeActionFromQueue()
        });
    }

}
