// pixi-core-framework/src/core/baseView/components/pixiProjection/BaseSpriteProjection2d.ts
import 'pixi-projection/dist/pixi-projection-spine';
import Sprite2d = PIXI.projection.Sprite2d;
import {TextureHolder} from "../../../utils/layouts/TextureHolder";
import {IBaseSpriteConfig} from "../../interfaces/IBaseSpriteConfig";
import {BaseViewComponentProjection2d} from "./BaseViewComponentProjection2d";
import Projection2d = PIXI.projection.Projection2d;
import {TransformBase} from "pixi.js";

export class BaseSpriteProjection2d extends BaseViewComponentProjection2d {

    protected _sourceData: IBaseSpriteConfig;
    protected initIsComplete: boolean = false;
    protected interactiveObject: boolean = false;
    protected _sprite2d:Sprite2d;

    public init(): void {
        if (this.initIsComplete) {
            return;
        }
        this._sprite2d = new Sprite2d(TextureHolder.getTextureByName(this._sourceData.image));
        this.addChild(this._sprite2d);
        this.initIsComplete = true;
        if (this._sourceData.width) {
            this.width = this._sourceData.width;
        }
        if (this._sourceData.height) {
            this.height = this._sourceData.height;
        }
        if (this._sourceData.angle !== undefined) {
            this.rotation = this._sourceData.angle * (Math.PI / 180);
        }
        if (this._sourceData.interactive !== undefined) {
            this.interactive = this._sourceData.interactive;
        }
    }

    get sourceData(): IBaseSpriteConfig {
        return this._sourceData;
    }

    get sprite2d():Sprite2d {
        return this._sprite2d;
    }
}
